crates/typst-syntax/src/parser.rs

```diff
@@ -455,84 +472,96 @@ fn math_args(p: &mut Parser) {
-                namable = true;
-                if named.is_some() {
-                    array = p.marker();
-                    named = None;
+            SyntaxKind::End | SyntaxKind::Dollar | SyntaxKind::RightParen => {
+                // Check if we have reached the end, after the last argument.
+                if has_arrays && positional {
+                    p.wrap(maybe_array_start, SyntaxKind::Array);
+                    // So we don't wrap as an array twice after the loop.
+                    array_last = false;
                 }
-                continue;
             }
-            _ => {}
-        }
-
-        if p.at_set(set::MATH_EXPR) {
-            math_expr(p);
-            count += 1;
-        } else {
-            p.unexpected();
+            _ => p.unexpected(),
         }
+    }
 
-        namable = false;
+    // Check if we need to wrap the preceding arguments in an array.
+    if array_last && maybe_array_start != p.marker() && has_arrays && positional {
+        p.wrap(maybe_array_start, SyntaxKind::Array);
     }
 
-    if arg != p.marker() {
-        maybe_wrap_in_math(p, arg, count, named);
-        if named.is_some() {
-            array = p.marker();
+    p.expect_closing_delimiter(m, SyntaxKind::RightParen);
+    p.wrap(m, SyntaxKind::Args);
+}
+
+/// Parses a single argument in a math argument list.
+///
+/// Returns whether the argument parsed was positional or not.
+fn math_arg<'s>(p: &mut Parser<'s>, seen: &mut HashSet<&'s str>) -> bool {
+    let m = p.marker();
+    let start = p.current_start();
+
+    if p.at(SyntaxKind::Dot) {
+        // Parses a spread argument: `..args`.
+        if let Some(spread) = p.lexer.is_math_spread_arg(start) {
+            p.token.node = spread;
+            p.eat();
+            math_expr(p);
+            p.wrap(m, SyntaxKind::Spread);
+            return true;
         }
     }
 
-    if has_arrays && array != p.marker() {
-        p.wrap(array, SyntaxKind::Array);
+    let mut positional = true;
+    if p.at_set(syntax_set!(Text, MathIdent, Underscore)) {
+        // Parses a named argument: `thickness: #12pt`.
+        if let Some(named) = p.lexer.is_math_named_arg(start) {
+            p.token.node = named;
+            let text = p.current_text();
+            p.eat();
+            p.convert_and_eat(SyntaxKind::Colon);
+            if !seen.insert(text) {
+                p[m].convert_to_error(eco_format!("duplicate argument: {text}"));
+            }
+            positional = false;
+        }
     }
 
-    if p.at(SyntaxKind::Text) && p.current_text() == ")" {
-        p.convert_and_eat(SyntaxKind::RightParen);
-    } else {
-        p.expected("closing paren");
-        p.balanced = false;
-    }
+    // Parses a normal position argument.
+    let arg = p.marker();
+    let count = math_exprs(p, syntax_set!(End, Comma, Semicolon, RightParen));
+    let named = if positional { None } else { Some(m) };
+    maybe_wrap_in_math(p, arg, count, named);
```

Comment:
I feel like `maybe_wrap_in_math` doesn't really pull its weight as an abstraction anymore. Perhaps it should be inlined into `fn arg`?